### PR TITLE
Show build duration and upload size for each layer

### PR DESCRIPTION
**Pull request by vrutkovs**
This also prints a total duration and upload size to stderr

### CODE REVIEW

File: graph.py [hunk 1]
```diff
@@ -17,12 +39,105 @@ def __init__(self, builds):
...
+            nvr = None
+            parent_id = None
+            if layer_id in self.known_pulp_layers.keys():
+                nvr = self.known_pulp_layers[layer_id]
+                size = self.found_nvr_sizes[nvr]
+            else:
+                pulp_url = pulp_repo_url + '/%s/json' % layer_id
+                r = requests.get(pulp_url, verify=False)
+                r.raise_for_status()
+                build_json = r.json()
```
Comment (twaugh) [inline graph.py]: A nit-pick, but this isn't really build JSON. It's layer JSON.

File: graph.py [hunk 2]
```diff
@@ -17,12 +39,105 @@ def __init__(self, builds):
...
+        except Exception:
```
Comment (twaugh) [inline graph.py]: Worth warning here?

File: graph.py [hunk 3]
```diff
@@ -17,12 +39,105 @@ def __init__(self, builds):
...
+        if image_id in self.known_pulp_layers.keys():
+            size = self.found_nvr_sizes[expected_nvr]
+            self.pulp_upload_size[image_id] = size
+            sys.stderr.write("  found cached size for %s: %s\n" % (image_id, size))
+            return size
+
+        full_image_name = unique_repos[0]
+        image_name = '-'.join(strip_registry_from_image(full_image_name).split('/'))
+        pulp_repo_url = 'https://brew-pulp-docker01.web.prod.ext.phx2.redhat.com/' + \
+                        'pulp/docker/v1/redhat-%s' % image_name.split(':')[0]
```
Comment (twaugh) [inline graph.py]: This needs to go in config (or an environment variable) or something.

File: graph.py [hunk 4]
```diff
@@ -17,12 +39,105 @@ def __init__(self, builds):
...
+                r = requests.get(pulp_url, verify=False)
+                r.raise_for_status()
+                build_json = r.json()
+        except Exception:
+            return (None, None, None)
+
+        size = build_json.get('Size', 0)
+        parent_id = build_json.get('parent_id', None)
+        labels = build_json['config'].get('Labels', {'Name': '', 'Version': '', 'Release': ''})
+        nvr = '%s-%s-%s' % (labels['Name'], labels['Version'], labels['Release'])
```
Comment (twaugh) [inline graph.py]: Another nitpick, but this isn't a Koji build name (that would be using the BZComponent label), and it isn't really a docker image name (that would be "%s:%s-%s"). I think it's meant to be a docker image name, so technically it should be $Name:$Version-$Release.

File: graph.py [hunk 5]
```diff
@@ -2,13 +2,37 @@
 import json
 from osbs.utils import strip_registry_from_image
 import sys
+import datetime
+import requests
+from requests.packages.urllib3.exceptions import InsecureRequestWarning
+requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
+import re
+
+BASE_PULP_URL = "https://brew-pulp-docker01.web.prod.ext.phx2.redhat.com/"
```
Comment (twaugh) [inline graph.py]: We don't want hostnames in here.

File: graph.py [hunk 6]
```diff
@@ -96,11 +293,19 @@ def run(inputfile=None):
...
     tree = BuildTree(builds)
     tree.trim_excess_tags()
-    print(tree.as_graph_easy_txt(include_datestamp=True))
+    print(tree.as_graph_easy_txt(
+          include_datestamp=True, include_duration=True, include_upload=True))
+
+    (total_duration, total_upload_size, total_layers_size) = tree.calculate_totals()
 
+    sys.stderr.write("Total duration: %s\n" % str(datetime.timedelta(seconds=total_duration)))
+    sys.stderr.write("Total upload size: %s (%s to Pulp and %s to Brew)\n" % (
```
Comment (twaugh) [inline graph.py]: s/Brew/Koji/ to make this applicable to Fedora as well.

File: graph.py [hunk 7]
```diff
@@ -17,12 +41,110 @@ def __init__(self, builds):
...
+        )
+        if image_id in self.known_pulp_layers.keys():
+            size = self.found_nvr_sizes[expected_nvr]
+            self.pulp_upload_size[image_id] = size
+            sys.stderr.write("  found cached size for %s: %s\n" % (image_id, size))
+            return size
+
+        full_image_name = unique_repos[0]
+        image_name = '-'.join(strip_registry_from_image(full_image_name).split('/'))
+        pulp_repo_url = '%s/pulp/docker/v1/redhat-%s' % (BASE_PULP_URL, image_name.split(':')[0])
```
Comment (twaugh) [inline graph.py]: You know, I was sort of expecting this to be talking to Crane instead of Pulp, i.e. looking at the [Docker Registry HTTP V1 API](https://docs.docker.com/v1.6/reference/api/registry_api/). Doing that would also avoid having to translation Docker image names into Pulp registry IDs (the "redhat-" prefix).

Comment (vrutkovs) [inline graph.py]: I was writing a crane version, but somehow the same V1 API calls gave HTTP 404 here.

I'll give it another look, as it should work much faster

Comment (twaugh) [inline graph.py]: That can happen if the image was removed from Pulp since being built (e.g. it was a beta image no longer in beta, or dropped). I think in that case there should just be a warning.

File: graph.py [hunk 8]
```diff
@@ -2,13 +2,36 @@
...
+    for unit in ['', 'K', 'M', 'G', 'T', 'P', 'E', 'Z']:
+        if abs(num) < 1024.0:
+            return "%3.1f%s%s" % (num, unit, suffix)
+        num /= 1024.0
+    return "%.1f%s%s" % (num, 'Yi', suffix)
 
 
 class BuildTree(object):
-    def __init__(self, builds):
+    def __init__(self, builds, pulp_base_url):
```
Comment (twaugh) [inline graph.py:21]: Can the tool be run without collecting this information? It would be useful to be able to analyse build JSON "offline".

Comment (vrutkovs) [inline graph.py:21]: Yes, but the results would be incorrect - since some builds have compressed size stored and some don't.

I'll probably update the script for OSD - both image-id and size should be correct there

File: graph.py [hunk 9]
```diff
@@ -75,32 +267,56 @@ def formatwhen(name):
...
+        sizeof_fmt(total_upload_size)))
 
 if __name__ == '__main__':
-    try:
-        run(sys.argv[1])
-    except IndexError:
-        run()
+    args = []
+    if len(sys.argv) > 1:
+        args = [sys.argv[1], sys.argv[2]]
```
Comment (twaugh) [inline graph.py]: This doesn't look quite right. If `len(sys.argv)==2`, won't this give an `IndexError`?

File: zabbix_metrics_watcher.py [hunk 10]
```diff
@@ -70,7 +70,7 @@ def durations(self):
         except:
             return {}
 
-    def send_zabbix_notification(self, zabbix_host, osbs_master, concurrent_builds):
+    def send_zabbix_notification(self, zabbix_host, osbs_master, concurrent_builds, builds_in_queue):
```
Comment (twaugh) [inline zabbix_metrics_watcher.py]: Should this be in a separate pull request?

File: graph.py (hunk 9 above)
Comment (vrutkovs) [inline graph.py]: Right, that's a dirty hack to avoid catching `IndexError` in `run`, I'm gonna rewrite that

File: zabbix_metrics_watcher.py (hunk 10 above)
Comment (vrutkovs) [inline zabbix_metrics_watcher.py]: That's odd, I thought I've already pushed that to master. Right, I'll make another PR for builds in queue

File: graph.py (hunk 8 above)
Comment (twaugh) [inline graph.py:21]: I was thinking it could be that `pulp_base_url` is None and we don't try calculating any sizes.

Comment (vrutkovs) [inline graph.py:21]: Implemented - when `pulp_base_url` is not specified we use `tar_metadata` annotation

**Comment by twaugh** [comment]
The README.md file needs updating, but otherwise I think this is ready to merge.

**Comment by vrutkovs** [comment]
Updated README and squashed into one commit
